### ISSUE TITLE
Pattern %d-%m-%y matches 2016-11-30

### ISSUE THREAD

**Issue by ehuelsmann** [houseabsolute/DateTime-Format-Strptime#11]
Migrated from https://rt.cpan.org/Ticket/Display.html?id=119082 :

Quoting:

Hi,

Using v1.67, I'm iterating through these patterns:

   '%d-%m-%Y'
   '%d-%m-%y'
   '%F'

to match the input '2016-11-30' against. It's not clear from the documentation that it's correct for my input to match the second pattern. I know it's definitely *very* unintuitive that it did -- and it obviously invalidates my use case... 

But from the fact that the parsed date is 2030-11-16, I can only conclude that the second pattern is the one to match. (Also, in the code, I can't find that you're using string or line delimiters to make sure you're matching *exactly that pattern* (and nothing more)). Thanks for having a look at this!

Regards,

Erik.

**Comment by chansen** [comment]
Thank you for the report! I concur, this is a bug.

`%d` should only accept a valid day of the month [1, 31], optionally zero padded or space padded: 
`[’1’, ’9’]` /  ` [’ 1’, ’ 9’]` /  `[’01’, ’09’]` / `[’10’, ’31’]`.

`%m` should only accept a valid month of the year [1, 12], optionally zero padded or space padded:
`[’1’, ’9’]` /  ` [’ 1’, ’ 9’]` /  `[’01’, ’09’]` / `[’10’, ’12’]`.

`%y` should accept a year of the century [0, 99], requiring zero padding for a year < 10:
`[’00’, ’99’]`

`%Y` should accept a year of the millennia, requiring zero padding for a year < 1000:
`[’0001’, ’9999’]`

--
chansen

**Comment by autarch** [comment]
While @chansen is right that the matching should be stricter for days & months, that's not the problem here. Even if I add that it still fails. The problem is that ultimately we end up with something like this:

```
my $day = qr/(?:[12][0-9]|3[01]|[0 ]?[1-9])/;
my $month = qr/(?:1[0-2]|[0 ]?[0-9])/;
my $regex = qr/(\A|\s|\b)*$day-$month-\d{1,4}/;
```

It's the `*` after the `(\A|\s|\b)` that's the real culprit!

**Comment by ehuelsmann** [comment]
@autarch that seems like a trivial fix. If you want, I can issue a PR for it.

**Comment by autarch** [comment]
@ehuelsmann - Thanks, I actually just merged a branch where I fixed it. I plan to have a new release out shortly.

**Comment by autarch** [comment]
Fixed in v1.69
